runtime/block_reduction.cu

```diff
@@ -111,3 +111,154 @@ __device__ void blockReduce(
       read_write_pred,
       init_val);
 }
+
+// Each thread in the iteration dimension processes N elements
+// Typical usage is in outer reduction where the iteration dimension
+// is parallelized by vectorized loads, bidmx. The reduction dimension
+// is parallelized by bdimy. This function works as follows:
+// (1) Each thread vectorized loads N elements from input register array to
+// smem. (2) do N * bdimx parallel reductions in smem.
+
+// TODO: merge `blockIterGroupedReduce` with `blockReduce`
+// (1) for-loops are fully unrolled should not cause overhead for `blockReduce`
+// (2) used in gridReduce, needs to change correspodning gridReduce function
+template <
+    bool X_REDUCE,
+    bool Y_REDUCE,
+    bool Z_REDUCE,
+    bool Aligned,
+    int N, // Number of elements per input array
+    typename T,
+    typename Func>
+__device__ void blockIterGroupedReduce(
+    T out[N],
+    const T inp_val[N],
+    Func reduction_op,
+    T* shared_mem,
+    bool read_pred,
+    bool write_pred,
+    T init_val) {
```

Comment:
Have you confirmed, for example, there's no bank conflicts? The store looks suspicious to me...